diff --git a/cartpole/ppo/tparallel.py b/cartpole/ppo/tparallel.py
index e00ed09..8a4c8e1 100644
--- a/cartpole/ppo/tparallel.py
+++ b/cartpole/ppo/tparallel.py
@@ -395,8 +395,9 @@ class Simulation:
             wandb.log(
                 {"Sampling Rollout time per step": wandb.plot.line(table_rollout, "rollouts", "time per rollout", title="Sampling time per Step")}
             )
-
             print(f"Global step value: {global_step}")
+
+
             self.writer.add_scalar('Average Value post rollouts amongst envs', scalar_value=self.value_buffer.mean(), global_step=global_step)
             self.writer.add_scalar('Average Returns post rollout amongst envs', scalar_value=self.reward_buffer.mean(), global_step=global_step)
 
diff --git a/cartpole/ppo/wandb/debug-internal.log b/cartpole/ppo/wandb/debug-internal.log
index 1736f7b..131ec98 120000
--- a/cartpole/ppo/wandb/debug-internal.log
+++ b/cartpole/ppo/wandb/debug-internal.log
@@ -1 +1 @@
-run-20231010_125613-v482y7x0/logs/debug-internal.log
\ No newline at end of file
+run-20231010_131843-9iaeb0um/logs/debug-internal.log
\ No newline at end of file
diff --git a/cartpole/ppo/wandb/debug.log b/cartpole/ppo/wandb/debug.log
index efc8bf7..fee9ec5 120000
--- a/cartpole/ppo/wandb/debug.log
+++ b/cartpole/ppo/wandb/debug.log
@@ -1 +1 @@
-run-20231010_125613-v482y7x0/logs/debug.log
\ No newline at end of file
+run-20231010_131843-9iaeb0um/logs/debug.log
\ No newline at end of file
diff --git a/cartpole/ppo/wandb/latest-run b/cartpole/ppo/wandb/latest-run
index 1b1db32..ed36bdc 120000
--- a/cartpole/ppo/wandb/latest-run
+++ b/cartpole/ppo/wandb/latest-run
@@ -1 +1 @@
-run-20231010_125613-v482y7x0
\ No newline at end of file
+run-20231010_131843-9iaeb0um
\ No newline at end of file
